### PR TITLE
types: Fix insert error when column length is zero

### PR DESCRIPTION
**Pull request by jackysp**
Fix insert error in #3832 
@XuHuaiyu @zz-jason PTAL

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
LGTM 
ptal @zz-jason

**Comment by jackysp** [comment]
PTAL @XuHuaiyu

File: util/types/datum.go [hunk 1]
```diff
@@ -848,7 +848,13 @@ func ProduceStrWithSpecifiedTp(s string, tp *FieldType, sc *variable.StatementCo
...
 				}
 			}
-			if truncateLen > 0 {
+			// 1. If len(s) is 0 and flen is 0, truncateLen will be 0, don't truncate s.
+			//    CREATE TABLE t (a char(0));
+			//    INSERT INTO t VALUES (``);
+			// 2. If len(s) is 10 and flen is 0, truncateLen will be 0 too, but we still need to truncate s.
+			//    SELECT 1, CAST(1234 AS CHAR(0));
+			// So truncateLen is not a suitable variable to determine to do truncate or not.
+			if len([]rune(s)) > flen {
```
Comment (XuHuaiyu) [inline util/types/datum.go]: can we put this check in line 841 ?

File: session_test.go [hunk 2]
```diff
@@ -2351,6 +2351,17 @@ func (s *testSessionSuite) TestSpecifyIndexPrefixLength(c *C) {
 	se.Close()
 }
 
+func (s *testSessionSuite) TestColumnLength(c *C) {
```
Comment (XuHuaiyu) [inline session_test.go]: we can put this test in `executor/write_test.go: TestInsert`,
and comment which issue is this case for.

Comment (jackysp) [inline session_test.go]: ok

File: util/types/datum.go (hunk 1 above)
Comment (jackysp) [inline util/types/datum.go]: under line 841 would be better.

File: expression/builtin_cast_test.go [hunk 3]
```diff
@@ -81,6 +81,15 @@ func (s *testEvaluatorSuite) TestCast(c *C) {
 		res, err = f.Eval(nil)
 		c.Assert(err, IsNil)
 	}
+
+	// cast(1234 as char(0))
+	tp.Flen = 0
+	tp.Charset = charset.CharsetUTF8
+	f = NewCastFunc(tp, &Constant{Value: types.NewDatum(1234), RetType: types.NewFieldType(mysql.TypeString)}, ctx)
+	res, err = f.Eval(nil)
+	c.Assert(err, IsNil)
```
Comment (shenli) [inline expression/builtin_cast_test.go]: There will be a warning in MySQL:
mysql> SELECT 1, CAST(1234 AS CHAR(0));
+---+-----------------------+
| 1 | CAST(1234 AS CHAR(0)) |
+---+-----------------------+
| 1 |                       |
+---+-----------------------+
1 row in set, 1 warning (0.00 sec)

Comment (shenli) [inline expression/builtin_cast_test.go]: I tried on this branch:
mysql>  SELECT 1, CAST(1234 AS CHAR(0));
+------+-----------------------+
| 1    | CAST(1234 AS CHAR(0)) |
+------+-----------------------+
|    1 |                       |
+------+-----------------------+
1 row in set (0.00 sec)

Comment (jackysp) [inline expression/builtin_cast_test.go]: The missing warning issue is also on master branch. I'll check it out.

Comment (jackysp) [inline expression/builtin_cast_test.go]: Truncate column should cause a warning in selection of MySQL, no matter strict or non-strict sql mode.

File: executor/prepared.go [hunk 4]
```diff
@@ -347,6 +347,10 @@ func ResetStmtCtx(ctx context.Context, s ast.StmtNode) {
 	case *ast.LoadDataStmt:
 		sc.IgnoreTruncate = false
 		sc.TruncateAsWarning = !sessVars.StrictSQLMode
+	case *ast.SelectStmt:
+		// Return warning for truncate error in selection.
+		sc.IgnoreTruncate = false
+		sc.TruncateAsWarning = true
```
Comment (shenli) [inline executor/prepared.go]: This is related with sessVars.StrictSQLMode.

**Comment by shenli** [comment]
LGTM

File: executor/prepared.go (hunk 4 above)
Comment (zimulala) [inline executor/prepared.go]: @jackysp

Comment (jackysp) [inline executor/prepared.go]: Selection in mysql always return a warning of truncate, no matter strict or non-strict sql mode,

**Pull request merged**